### ISSUE TITLE
early_stopping_rounds?

### ISSUE THREAD

**Issue by Moelf** [dmlc/XGBoost.jl#158]
I don't see it being explicitly tested and supported but if it's part of C++ library it should just work?

**Comment by Moelf** [comment]
![image](https://user-images.githubusercontent.com/5306213/211701653-6b7c4390-d3a1-4590-9465-6d61c8bef48c.png)

:(

**Comment by ExpandingMan** [comment]
I don't see that parameter listed [here](https://xgboost.readthedocs.io/en/stable/parameter.html)... what makes you think it exists?

Also, I'm not sure what the difference is supposed to be between this and `num_rounds`.

**Comment by Moelf** [comment]
`num_rounds` is the maximum rounds in some sense, early_stopping_rounds is the number of rounds without improvement for the fit to halt

**Comment by Moelf** [comment]
btw related question, is it possible to error when unsupported keywords gets passed to XGBoost? for example the `early_stopping_rounds`, and because it didn't error, I keep worrying about if I spelled `colsample_bytree` correctly

**Comment by ExpandingMan** [comment]
> btw related question, is it possible to error when unsupported keywords gets passed to XGBoost?

I would *really* like to, this has been an annoyance of mine as well.  I've also been burned by this many times. 
 Unfortunately right now there are no good options: the C API does not return an error code for invalid parameters, and the only way to retrieve a list is from the documentation or source code (i.e. there is no way to incorporate it to XGBoost.jl without a maintenance nightmare that will cause more harm than good).

I just opened [an issue](https://github.com/dmlc/xgboost/issues/8659) for this.

**Comment by bobaronoff** [comment]
From past work in R, early stopping is a convenience feature provided by the package as opposed to a libxgboost function. The feature relies on capturing the evaluation log and applying a criteria of when to stop.  The optimum num_round value can change with other parameters and it is difficult to tune num_round parameter independent of others.  There are implications in how one does a grid search.  

If it's important enough to a workflow, it is possible for one to write their own early stopping function. In order to plot learning curves from cross validation, without duplicating the metrics already available in libxgboost, it was necessary to 'intercept' and parse the evaluation log.  I posted my modest attempt(i.e. hack) in #148.  It has been working quite well although am sure it could be improved.  All are welcome to use and adapt to their needs.  For early stopping one would need to modify and parse each line of the log as it is recorded; the posted function parses all the lines together after the run.

Using this approach I've been able to create a function to duplicate the cv learning curves one is accustomed to in R (if that is one's workflow). An example ....

![fram2_0a](https://user-images.githubusercontent.com/5597810/213497753-ba816f0a-644b-4de7-9c5d-70bef22448b5.svg)

**Comment by Moelf** [comment]
yeah, well, we have to do this externally in Julia because we can't pass a callable to `libxgboost` for evaluation set evals

**Comment by ExpandingMan** [comment]
This isn't *that* hard to do with current XGBoost.jl, but yes, it is annoying that one can't access the evaluation metrics it's already computing via `watchlist` and instead has to compute it separately.

I'd be happy to help get through a PR that makes this more ergonomic.  I'm open to replacing the `watchlist` mechanism completely, though that might be pretty tough without breaking changes.

**Comment by bobaronoff** [comment]
Not sure I understand.  I am able to set the evaluation metrics with the eval_metric keyword parameter.  The results show up in the log and are parse-'able.  It does require a bit of an end-around by writing a personal version of one or two of the functions in booster.jl but the crux of it is only a handful of lines.

**Comment by bobaronoff** [comment]
Recopied from #148.  This is my function to capture the logs.

```
function xgboost_log(traindm::DMatrix, a...;
                        testdm::Any=[] ,
                        num_round::Integer=10,
                        kw...
                    )
    
    Xy = XGBoost.DMatrix(traindm)
    b = XGBoost.Booster(Xy; kw...)
    update_feature_names::Bool=false
    if typeof(testdm)==DMatrix
        watchlist=Dict("train"=>traindm , "test"=>testdm)
    else
        watchlist=Dict("train"=>traindm)
    end
    names = collect(Iterators.map(string, keys(watchlist)))
    watch = collect(Iterators.map(x -> x.handle, values(watchlist)))
    thelog = Vector{String}(undef,0)
    for j in 1:num_round
        XGBoost.xgbcall(XGBoost.XGBoosterUpdateOneIter, b.handle, j, Xy.handle)
        o = Ref{Ptr{Int8}}()
        XGBoost.xgbcall(XGBoost.XGBoosterEvalOneIter, b.handle, j, watch, names, length(watch), o)
        push!(thelog,unsafe_string(o[]))
        XGBoost._maybe_update_feature_names!(b, Xy, update_feature_names)
    end
    return (booster=b , log=parsethelog(thelog))
end
'''

**Comment by ExpandingMan** [comment]
I'm not enthused about the prospect of opening the pandora's box of parsing log output, it's bad enough that we are already trying to do this to identify warning messages, in my opinion it would be much nicer to just have it run evaluation normally so the whole thing is accessible and controllable.  That said, I wouldn't stand in the way of it, particularly if you left it flexible enough to leave it open for later improvements, it could be replaced with a non-parsing version at some point in the future.

**Comment by Moelf** [comment]
actually
>it's bad enough that we are already trying to do this to identify warning messages,

I'm not sure if it's for or against the idea to parsing logs to get loss during training, I mean given we already do this we might as well utilize existing infrastructure :)

**Comment by ExpandingMan** [comment]
> I'm not sure if it's for or against the idea to parsing logs to get loss during training, I mean given we already do this we might as well utilize existing infrastructure :)

It's not infrastructure, it's one badly-written regex, and it's done because there is no alternative.  For evaluations there is an alternative, there is no clear reason why this needs to be done by the `watchlist` arg, we already have plenty of library functions available for doing it.

Again, I'm not going to stand in the way of a PR that provides parsing of the logs to extract evaluation results as a feature, but I think we should keep it flexible enough that it could be made more reasonable in the future.

**Comment by Moelf** [comment]
one reason to not do it in Julia would be performance, also keep in mind we'd have to some how evaluate it with the same metric as the training and handle GPU etc.

**Comment by ExpandingMan** [comment]
> one reason to not do it in Julia would be performance

Why should it be more expensive to run via a separate call to the `predict` function?  The overhead should be minuscule.  If that's not the case, something is very wrong.

> also keep in mind we'd have to some how evaluate it with the same metric as the training and handle GPU etc.

We already have calls to do all this stuff, it shouldn't be hard.  I don't think it should be more complicated than providing a function argument to `update!` and providing reasonable defaults.

**Comment by bobaronoff** [comment]
Actually, it was the need for performance that lead me to explore parsing the log.  One could argue that learning curves are old fashioned but that is how I was trained to build a gradient boost model, so for me learning curve it is. I did write a cross validation function based on predict and it works okay.  However, consider the task.  In a typical 10 fold learning curve one requires 20 calls to predict per round (10 for train and 10 for test).  A typical model has 500 rounds.  That's 10,000 calls to predict and 10,000 calls to a function that processes the return in to a metric.  Although predict allows one to specify a particular round the lower rounds are not cache'd.  Predicting round 100 drops through all 1-99 trees even though it went through the same trees when predicting for round 99.  For even a simple model this is a lot of stuff.  Pre-allocating space for the predictions and as efficient broadcasting as I could, the processing time was considerably improved but still a bit long.  Reducing the curve to every 10 rounds brought the processing time to what I considered appropriate.  I haven't figured out yet how to do this with MLJ.  I am sure it will be better at the 1,000 metric calculations but the 1,000 calls to libxgboost for prediction will be the same. In comparison, the time to parse an evaluation log is trivial.  After I re-wrote my learning curve function with data from the parsed log, the run times with evaluations every round are comparable to the 'predict' method every 10 rounds.  Also, I don't need to keep track of or code the default metrics.  For my needs this seemed the better solution.  I did observe that booster.jl grows every tree one round at a time ( a call to libxgboost for each round).  I was wondering/hoping if libxgboost relegated any cache space for prior levels predictions as that is needed to grow the next tree, but that is off topic.  My best effort at creating a 'predict' approach may not be the best metric but it is what I have; parsing the evaluation log won out for my purpose and level of skill.  This is all predicated on simple cross validation.  If one is in to nested cross validation, comparing different algorithms, then MLJ would seem to be the best ( if not only) way to go.

Aside from my bashfulness, the reason I hesitate to submit a PR is that I am next to useless when it comes to git and GitHub.  It's all I can do to keep my personal packages straight and not corrupt my Julia environment.  I am afraid if I fork and download, my laptop might blow up due to my ignorance.

All in all, XGBoost.jl is just fine.

**Comment by david-sun-1** [comment]
Hey all, I too have had a need for early stopping rounds.
I've implemented something that closely follow @bobaronoff 's solution above.

Understand that there may be better ways, but it's a start and want to work towards getting this functionality added.

https://github.com/dmlc/XGBoost.jl/pull/193

**Comment by Moelf** [comment]
btw, one problem with early_stopping_rounds now is it monitors everything in `watchlist`, but often you have both Training loss and Validation loss in `watchlist`...
